engine.py

```diff
@@ -3,243 +3,491 @@
+            else:
+                try:
+                    if self.board[r + 1][c] == "  ":  # if square is empty or square has opponent's piece
+                        moves.append(Move((r, c), (r + 1, c), self.board))  # create a move object and append to moves
+                except IndexError:
+                    pass
+
+            try:
+                ''' so it doesnt crash when a piece reaches the end of the board
+                    ideally the pawn should promote to either q, r, b, n
+                    but i haven't been able to write the code yet'''
+                # same capture moves as in light
+                if c == 0 and self.board[r + 1][c + 1][1] == "l":
+                    moves.append(Move((r, c), (r + 1, c + 1), self.board))
+                elif c == 7 and self.board[r + 1][c - 1][1] == "l":
+                    moves.append(Move((r, c), (r + 1, c - 1), self.board))
+                elif c != 0 and c != 7:
+                    for i in range(-1, 2, 2):
+                        if self.board[r + 1][c + i][1] == "l":
+                            moves.append(Move((r, c), (r + 1, c + i), self.board))
+            except IndexError:
+                pass
 
+ 
 
-	def make_move(self, move):
-		"""
-			moves pieces on the board
-		"""
-		self.board[move.start_row][move.start_col] = "  "
-		self.board[move.end_row][move.end_col] = move.piece_moved
-		self.move_log.append(move) # log move
-		self.light_to_move = not self.light_to_move # next player to move
-
-
-	def undo_move(self, look_ahead_mode = False):
-		"""
-			undoes last move
-		"""
-		if self.move_log:
-			last_move = self.move_log.pop()
-			self.board[last_move.start_row][last_move.start_col] = last_move.piece_moved
-			self.board[last_move.end_row][last_move.end_col] = last_move.piece_captured
-			self.light_to_move = not self.light_to_move
-
-			print("undoing ->", last_move.get_chess_notation())
-		else:
-			print("All undone!")
-
-
-	def get_valid_moves(self):
-		return self.get_possible_moves()
-
-
-	def get_possible_moves(self):
-
-		moves = []
-
-		turn = "l" if self.light_to_move else "d"
-
-		for i in range(len(self.board)):
-			for j in range(len(self.board[i])):
-				if self.board[i][j][1] == turn:
-					self.move_piece[self.board[i][j][0]](i, j, moves)
-
-		return moves, turn
+    def get_bishop_moves(self, r, c, moves):
+
+        """
+            calculates all possible bishop moves for a given colour (light or dark)
+            and appends them to a list
+
+            input parameters:
+            r     --> starting row (int)
+            c     --> starting column (int)
+            moves --> posiible moves container (list)
+
+            return parameter(s):
+            None
+        """
+
+        ##TODO
+        pass
+
+
+    def get_knight_moves(self, r, c, moves):
+        ##TODO
+        pass
+
+
+    def get_king_moves(self, r, c, moves):
```

Comment:
I'm positive you were not given get_king_moves() task. This should be left empty for your teammates

Reply:
I'm sorry, I added them so I can test the rook moves properly I will change them and push again